include/pagmo/algorithms/gaco.hpp

```diff
@@ -0,0 +1,988 @@
+            alpha = 1.0 - 1.0 / (2.0 * std::sqrt(diff / m_res));
+
+        } else if (fitness > m_oracle && m_res > diff) {
+            alpha = 1.0 / 2.0 * std::sqrt(diff / m_res);
+        }
+
+        // I can now compute the penalty function value
+        if (fitness > m_oracle || m_res > 0.) {
+            penalty = alpha * diff + (1 - alpha) * m_res;
+
+        } else if (fitness <= m_oracle && m_res == 0.) {
+            penalty = -diff;
+        }
+
+        return penalty;
+    }
+
+    /**
+     * Function which updates the solution archive, if better solutions are found
+     *
+     * @param[in] pop Population: the current population is passed
+     * @param[in] sorted_vector Stored penalty vector: the vector in which the penalties of the current population are
+     * stored from the best to the worst is passed
+     * @param[in] sorted_list Positions of stored penalties: this represents the positions of the individuals wrt their
+     * penalties as they are stored in the stored_vector
+     * @param[in] sol_archive Solution archive: the solution archive is useful for retrieving the current
+     * individuals (which will be the means of the new pdf)
+     */
+
+    void update_sol_archive(const population &pop, vector_double &sorted_vector,
+                            std::vector<decltype(sorted_vector.size())> &sorted_list,
+                            std::vector<vector_double> &sol_archive) const
+    {
+
+        auto variables = pop.get_x();
+        auto fitness = pop.get_f();
+
+        // This part can only be used for single obj:
+        vector_double fitness_sol_arch(fitness[0].size());
+
+        for (decltype(fitness[0].size()) j = 0u; j < fitness[0].size(); ++j) {
+            fitness_sol_arch[j] = sol_archive[0][1 + variables[0].size() + j];
+        }
+
+        std::vector<vector_double> old_archive(sol_archive);
+        std::vector<vector_double> temporary_archive(sol_archive);
+        vector_double temporary_penalty(m_ker);
+
+        // I now re-order the variables and objective vectors (remember that the objective vector also contains the eq
+        // and ineq constraints). This is done only if at least the best solution of the current population is better
+        // than the worst individual stored in the solution archive
+        if (sorted_vector[0] < old_archive[m_ker - 1][0]) {
+
+            // I reset the impstop counter since the solution archive is updated:
+            m_n_impstop = 1;
+            // I save the variables, objectives, eq. constraints, ineq. constraints and penalties in three different
+            // vectors:
+            for (decltype(m_ker) i = 0u; i < m_ker; ++i) {
+                variables[i] = pop.get_x()[sorted_list[i]];
+                fitness[i] = pop.get_f()[sorted_list[i]];
+                temporary_penalty[i] = sol_archive[i][0];
+            }
+            std::vector<unsigned> saved_value_position;
```

Comment:
Same general comments as above, would be good to avoid re-creating vectors at every loop iteration.

Reply:
@bluescarni 
I have corrected this aspect in all the abovementioned vectors, thanks. However, this `saved_value_position` is only generated once per generation when the function is called. Shall I declare it in the `main` (outside the generations loop) and pass it as an input for the function every time?

Reply:
I think it's fine to leave it there for the time being.